tvm-tac/src/main/kotlin/org/ton/tac/Stack.kt

```diff
@@ -510,74 +533,290 @@ class Stack(
+
+        val tupleVar =
+            TacTupleValue(
+                name = "t_${ctx.nextVarId()}",
+                elements = tupleElements,
+            )
+        registerState.tupleRegistry[tupleVar.name] = tupleElements
+
+        tupleElements.forEach { element ->
+            inputs.add(element.name to element)
+        }
+
+        metaObjects.add(input.name to tupleVar)
+    }
+
+    private fun handleSimpleInput(
+        input: TvmStackEntryDescription,
+        continuationMap: MutableMap<String, ContinuationId>,
+        registerState: RegisterState,
+        inputs: MutableList<Pair<String, TacStackValue>>,
+    ) {
+        val specInput = input as TvmSimpleStackEntryDescription
+        val specValueTypes = specInput.valueTypes
+
+        val poppedValue = popWithTypeCheck(expectedTypes = specValueTypes)
+        val valueToStore =
+            if (specValueTypes.contains(TvmType.TUPLE)) {
+                if (poppedValue is TacTupleValue) {
+                    poppedValue
+                } else {
+                    var tmp: String? = null
+                    for (x in registerState.tupleRegistry.keys) {
+                        if (x.startsWith("global")) tmp = x
+                    }
+                    val result =
+                        TacTupleValue(
+                            name = poppedValue.name,
+                            elements = registerState.tupleRegistry[tmp] ?: emptyList(),
+                        )
+
+                    if (tmp != null) {
+                        registerState.tupleRegistry.remove(tmp)
+                    }
+                    result
+                }
+            } else {
+                poppedValue
+            }
+
+        if (valueToStore is ContinuationValue) {
+            continuationMap[input.name] = valueToStore.continuationRef
+        }
+
+        inputs.add(input.name to valueToStore)
+    }
+
+    private fun handleSimpleOutput(
+        output: TvmStackEntryDescription,
+        ctx: TacGenerationContext<*>,
+        contRef: Int?,
+        outputs: MutableList<TacStackValue>,
+        instruction: TvmRealInst,
+        metaObjects: MutableList<Pair<String, TacStackValue>>,
+        value: Int?,
+        inputs: MutableList<Pair<String, TacStackValue>>,
+        registerState: RegisterState,
+    ) {
+        val specOutput = output as TvmSimpleStackEntryDescription
+        val specName = specOutput.name
+        val specValueTypes = specOutput.valueTypes
+        val name = "${specName}_${ctx.nextVarId()}"
+
+        val pushValue =
+            if (contRef != null) {
+                ContinuationValue(name, contRef).also {
+                    check(it.valueTypes == specValueTypes) {
+                        "Unexpected output value types for $output. Expected ${it.valueTypes}."
+                    }
+                }
+            } else {
+                if (specValueTypes.contains(TvmType.TUPLE) && instruction.mnemonic == "TPUSH") {
```

Comment:
I don't like that this is some special implementation for `TPUSH`. Can't you make some universal implementation?

Reply:
Or, if you want not to lose information in this specific case, you should probably move such special cases, for example, in a separate file